sublaunching/sublaunching_tutorial/README.md

```diff
@@ -0,0 +1,52 @@
+
+Note 2: The scripts in this repository require Python version 3.6 or above.
+
+## 1 - Preparing your SLiM file for command line configuration
+
+Just as you can use the defineConstant() function to... well... define constants within SLiM, you can use the command line "-d" argument to do so as well.
+
+For example invoking: ``slim -d X=0.0 -d Y=1.0 filename.slim`` will run SLiM with X in the namespace as 0 and Y as 1. This can be used in conjunction with other tools to run many SLiM simulations while varying parameters.
+
+However, you'll need to make sure not to define the same constants twice! Doing so results in an error. To define a constant "X" to have a default value within SLiM in such a way that it can also be defined via the command line without causing an error, use the following code:
+```
+if (!exists("X"))
+    defineConstant(X, 1.0);
+```
+However, if you plan to use lots of configurable constants, this may start to take up a lot of space, and a small helper function can clean this up. See the "minimal_gene_drive.slim" file for an example of how this is done.
+
+## 2 - Using bash to run SLiM
+
+Once your SLiM file is prepared, running it with varrying parameters using bash is easy. Just a simple list of the command line strings you want to run is sufficient, though you may also want to pipe the output of each simulation to its own file via "``>``" (e.g. "``command > output.txt``"). You can also run jobs in parallel very easily.
+
+The "array_without_python.sh" file in this folder demonstrates how to accomplish parallel running with bash. The file can be run via ``bash array_without_python.sh``.
+
+## 3 - Using Python to run SLiM
+
+You might find that you want to do some complex analyses on your data which are inconvenient to do from within SLiM itself. If so, you can capture the results of SLiM within Python using the subprocess library, and then use all of your favorite python libraries to analyze the output.
+
+For a complete though fairly minimal example of such a driver, see the "minimal_slim_driver.py" file.
+
+## 4 - Using Python and bash to run an array in parallel
+
+Running SLiM in parallel using bash and running SLiM with python can be combined. The "array_with_python.sh" script in this folder demonstrates this method, and can be run via ``bash array_with_python.sh``. This could also be combined with a script to generate a bash file: see "generate_large_local_array_run.py". This file canbe invoked via ``python generate_large_local_array_run.py > large_local_array_run.sh``. The file that is generated can be run via ``bash large_local_array_run.sh``. Warning: this will cause 120 SLiM processes to be started, and your computer will run as many of these in parallel as it is capable of. Your CPU fan will definitely spin up a bit.
```

Comment:
*can be

Re: "your computer will run as many of these in parallel as it is capable of", I think it would be more accurate to say: "your computer will run all of these simultaneously, even if you have fewer than 120 cores.  This may mean that memory usage is excessive – all 120 processes running simultaneously – and that performance suffers because multiple processes are contending for the same core.  Better is method #5 below, which uses a cluster job scheduler to run one job per core, automatically starting new jobs as the previous jobs finish.

Reply:
This is fair. Because I mostly work on gene drive stuff, where the number of genetic loci being simulated are very few, I don't normally worry that much about how much memory SLiM uses! However, I suspect that many folks who use SLiM don't have easy access to a computing cluster, and that a handy workflow for running moderately sized local arrays might be very useful.

Thus, I've done a minor tweak on generate_large_local_array_run.py in order to limit the number of SLiM procs that get spooled up simultaneously, in order to be more reasonable on memory. There are better, but more complex ways to do this, I suppose, but just for purposes of being an easy tutorial that folks can apply to their own projects, I think this is a fine balance.

Reply:
Oh, for sure – I'm not suggesting that you write a job scheduler or anything.  I just want the description in the text to be more explicit about what will happen, and the tradeoffs involved.  :->